File: Upcoder/Lập trình hướng đối tượng/4.Kế thừa/Lớp hình chữ nhật, hình vuông/HCNHV.cpp
#include<bits/stdc++.h>
using namespace std;
class HCN
{
    protected:
        double _dai;
        double _rong;
    public:
        HCN(double dai, double rong);
        double ChuVi();
        double DienTich();
};

class HV : public HCN
{
    public:
    HV(double canh) : HCN(canh, canh) {};
    double ChuVi();
};

int main()
{
    vector<HCN*> danhSachHinh;
    string loaiHinh;
    double maxChuVi = 0;

    cin >> loaiHinh;
    cin.ignore(); // Bỏ qua ký tự xuống dòng sau khi nhập loại hình

    string line;
    while (getline(cin, line) && !line.empty()) {
        double a, b;
        if (loaiHinh == "HCN") {
            if (sscanf(line.c_str(), "%lf %lf", &a, &b) == 2) {
                danhSachHinh.push_back(new HCN(a, b));
            }
        } else if (loaiHinh == "HV") {
            if (sscanf(line.c_str(), "%lf", &a) == 1) {
                danhSachHinh.push_back(new HV(a));
            }
        }
    }
    
    // Tìm chu vi lớn nhất
    for(auto h : danhSachHinh)
    {
        double Chuvi = h->ChuVi();
        if(Chuvi > maxChuVi)
            maxChuVi = Chuvi;
    }
    cout << maxChuVi;
    
    // Giải phóng bộ nhớ
    for (auto hinh : danhSachHinh) {
        delete hinh;
    }
}

HCN::HCN(double dai, double rong)
{
    _dai = dai;
    _rong = rong;
}

double HCN::DienTich() {
    return _dai * _rong;
}

double HCN::ChuVi() {
    return 2 * (_dai + _rong);
}

double HV::ChuVi() {
    return 4 * _dai;
}

File: Upcoder/Lập trình hướng đối tượng/4.Kế thừa/Lớp hình chữ nhật, hình vuông/HCNHV.cs
using System;
using System.Collections.Generic;
namespace Hinhhoc
{
    public class HinhChuNhat
    {
        protected double _chieudai;
        protected double _chieurong;

        public HinhChuNhat(double dai, double rong)
        {
            _chieudai = dai;
            _chieurong = rong;
        }

        public virtual double ChuVi()
        {
            return 2 * (_chieudai + _chieurong);
        }

        public double DienTichHCN()
        {
            return  _chieudai * _chieurong;
        }
    }

    public class HinhVuong:HinhChuNhat
    {
        public HinhVuong(double canh) : base(canh, canh) { }
        public override double ChuVi()
        {
            return 4 * _chieudai;
        }
    }
    public class Program
    {
        public static void Main(string[] args)
        {
            List<HinhChuNhat> danhSachHinh = new List<HinhChuNhat>();
            double maxchuvi = 0;
            
            string input = Console.ReadLine();;
            while(true)
            {
                string line = Console.ReadLine();
                if (string.IsNullOrEmpty(line)) break;  // Dừng khi gặp dòng trống

                string[] parts = line.Split();

                if(input == "HCN")
                {
                    double dai = double.Parse(parts[0]);
                    double rong = double.Parse(parts[1]);
                    danhSachHinh.Add(new HinhChuNhat(dai, rong));
                }
                else if(input == "HV")
                {
                    double canh = double.Parse(parts[0]);
                    danhSachHinh.Add(new HinhVuong(canh));
                }
            }
            // Tìm chu vi lớn nhất
            foreach (var hinh in danhSachHinh)
            {
                double chuVi = hinh.ChuVi();
                if (chuVi > maxchuvi)
                    maxchuvi = chuVi;
            }
            Console.Write(maxchuvi);
        }
    }
}